src/interfaces/qwrendererinterface.cpp

```diff
@@ -0,0 +1,366 @@
+}
+
+static bool read_pixels(wlr_renderer *handle,
+                        uint32_t fmt,
+                        uint32_t stride,
+                        uint32_t width,
+                        uint32_t height,
+                        uint32_t srcX,
+                        uint32_t srcY,
+                        uint32_t dstX,
+                        uint32_t dstY,
+                        void *data)
+{
+    QImage::Format format {toQImageFormat(fmt)};
+    if (format == QImage::Format::Format_Invalid) {
+        qFatal("%s", QString("bad format. [format]: %1").arg(fmt).toUtf8().data());
+    }
+
+    auto pixelFormat {QImage::toPixelFormat(format)};
+    auto pixelSize = pixelFormat.bitsPerPixel();
+    QImage image {static_cast<uchar *>(data) + dstY * stride + dstX * pixelSize,
+                 uint32ToPixelSize(width),
+                 uint32ToPixelSize(height),
+                 uint32ToPixelSize(stride),
+                 format};
+
+    const QRect srcRect {uint32ToPixelSize(srcX),
+                        uint32ToPixelSize(srcY),
+                        uint32ToPixelSize(width),
+                        uint32ToPixelSize(height)};
+
+    return interface(handle)->readPixels(srcRect, image);
+}
+
+static void destroy(wlr_renderer *handle)
+{
+    delete interface(handle);
+}
+
+static int get_drm_fd(wlr_renderer *handle)
+{
+    return interface(handle)->getDrmFd();
+}
+
+static uint32_t get_render_buffer_caps(wlr_renderer *handle)
+{
+    return interface(handle)->getRenderBufferCaps();
+}
+
+static wlr_texture *texture_from_buffer(wlr_renderer *handle, wlr_buffer *buffer)
+{
+    auto *texture = interface(handle)->textureFromBuffer(QWBuffer::from(buffer));
+    return texture ? texture->handle() : nullptr;
+}
+
+#if WLR_VERSION_MAJOR == 0 && WLR_VERSION_MINOR > 16
+static wlr_render_pass *begin_buffer_pass(wlr_renderer *handle, wlr_buffer *buffer) {
+    return interface(handle)->beginBufferPass(QWBuffer::from(buffer));
+}
+#endif
+} // namespace impl
+
+QWRendererInterface::~QWRendererInterface()
+{
+    free(handle());
+    delete impl();
+}
+
+QWRendererInterface *QWRendererInterface::get(QWRenderer *handle)
+{
+    return interface(handle->handle());
+}
+
+bool QWRendererInterface::bindBuffer(QWBuffer *)
+{
+    return false;
+}
+
+const QVector<uint32_t>* QWRendererInterface::getShmTextureFormats() const
+{
+    return nullptr;
+}
+
+const wlr_drm_format_set *QWRendererInterface::getDmabufTextureFormats() const
+{
+    return nullptr;
+}
+
+const wlr_drm_format_set *QWRendererInterface::getRenderFormats() const
+{
+    return nullptr;
+}
+
+uint32_t QWRendererInterface::preferredReadFormat() const
+{
+    return 0;
+}
+
+bool QWRendererInterface::readPixels(const QRect &, QImage &)
+{
+    return false;
+}
+
+int QWRendererInterface::getDrmFd() const
+{
+    return 0;
+}
+uint32_t QWRendererInterface::getRenderBufferCaps() const
+{
+    return 0;
+}
+
+#if WLR_VERSION_MAJOR == 0 && WLR_VERSION_MINOR <= 17
```

Comment:
Why is `<=17`? Is the function removed from wlroots 0.18.0 ? But  the wlroots 0.17.x is not released.